validator.go

```diff
@@ -0,0 +1,91 @@
+/**
+ *  Copyright (c) 2022  Comcast Cable Communications Management, LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+package wrp
+
+import (
+	"errors"
+)
+
+var (
+	ErrInvalidMsgTypeValidator = errors.New("invalid WRP message type validator")
+	ErrInvalidMsgType          = errors.New("invalid WRP message type")
+)
+
+// AlwaysInvalid doesn't validate anything about the message and always returns an error.
+var AlwaysInvalidMsg ValidatorFunc = func(m Message) error {
+	return ErrInvalidMsgType
+}
+
+// Validator is a WRP validator that allows access to the Validate function.
+type Validator interface {
+	Validate(m Message) error
+}
+
+// Validators is a WRP validator that ensures messages are valid based on
+// message type and each validator in the list.
+type Validators []Validator
+
+// ValidatorFunc is a WRP validator that takes messages and validates them
+// against functions.
+type ValidatorFunc func(Message) error
+
+// Validate executes its own ValidatorFunc receiver and returns the result.
+func (vf ValidatorFunc) Validate(m Message) error {
+	return vf(m)
+}
+
+// MsgTypeValidator is a WRP validator that validates based on message type
+// or using the defaultValidator if message type is unknown.
+type MsgTypeValidator struct {
+	m                map[MessageType]Validators
+	defaultValidator Validator
+}
+
+// Validate validates messages based on message type or using the defaultValidator
+// if message type is unknown.
+func (m MsgTypeValidator) Validate(msg Message) error {
+	vs, ok := m.m[msg.MessageType()]
+	if !ok {
```

Comment:
Rather than use `ok`, we can do a nil check.  This is the easiest way to also make sure we aren't trying to invoke on a `nil` interface:
```suggestion
	vs := m.m[msg.MessageType()]
	if vs == nil {
```
This works because the empty value of the type is returned when the key isn't set - for interfaces, the empty value is nil. ([playground](https://play.golang.com/p/oaZCdURarHk))

---
Some alternatives: 
- we can do a nil check of `vs` after this and error out.
- we can check all map values in the New function.